intellij/src/saros/intellij/eventhandler/filesystem/LocalFilesystemModificationHandler.java

```diff
@@ -367,35 +368,103 @@ private void generateResourceDeletionActivity(@NotNull VirtualFileEvent virtualF
       }
 
       return;
     }
 
     User user = session.getLocalUser();
 
-    IActivity activity;
+    Deque<IActivity> queuedDeletionActivities = new ConcurrentLinkedDeque<>();
 
-    if (deletedVirtualFile.isDirectory()) {
-      // TODO create deletion activities for child resources
-      // TODO clean up editor pool and annotations for child resources
-      activity = new FolderDeletedActivity(user, path);
+    VirtualFileFilter virtualFileFilter = getVirtualFileFilter(deletedFolder);
 
-    } else {
-      activity =
-          new FileActivity(
-              user, Type.REMOVED, FileActivity.Purpose.ACTIVITY, path, null, null, null);
+    /*
+     * Defines the actions executed on the base directory and every valid
+     * child resource (defined through the virtualFileFilter).
+     *
+     * Returns whether the content iteration should be continued.
+     */
+    ContentIterator contentIterator =
+        fileOrDir -> {
+          if (!fileOrDir.isDirectory()) {
+            generateFileDeletion(fileOrDir);
+
+            return true;
+          }
+
+          SPath folderPath = VirtualFileConverter.convertToSPath(project, fileOrDir);
+
+          if (folderPath == null) {
+            LOG.debug("Skipping deleted folder as no SPath could be obtained " + fileOrDir);
+
+            return true;
+          }
+
+          IActivity newFolderDeletedActivity = new FolderDeletedActivity(user, folderPath);
+
+          queuedDeletionActivities.addFirst(newFolderDeletedActivity);
+
+          return true;
+        };
 
-      editorManager.removeAllEditorsForPath(path);
+    /*
+     * Calls the above defined contentIterator on the base directory and
+     * every contained resource.
+     * Directories are only stepped into if they match the above defined
+     * virtualFileFilter. This also applies to the base directory.
+     */
+    VfsUtilCore.iterateChildrenRecursively(deletedFolder, virtualFileFilter, contentIterator);
+
+    while (!queuedDeletionActivities.isEmpty()) {
+      dispatchActivity(queuedDeletionActivities.pop());
+    }
+  }
+
+  /**
+   * Generated and dispatches a file deletion activity for the deleted file. Subsequently removes
+   * any editors for the file from the editor pool and drops any held annotations for the file.
+   *
+   * @param deletedFile the file that was deleted
+   */
+  private void generateFileDeletion(VirtualFile deletedFile) {
+    SPath path = VirtualFileConverter.convertToSPath(project, deletedFile);
+
+    if (path == null || !session.isShared(path.getResource())) {
+      if (LOG.isTraceEnabled()) {
+        LOG.trace("Ignoring non-shared file deletion: " + deletedFile);
+      }
 
-      annotationManager.removeAnnotations(path.getFile());
+      return;
     }
 
+    User user = session.getLocalUser();
+
+    IActivity activity =
+        new FileActivity(user, Type.REMOVED, FileActivity.Purpose.ACTIVITY, path, null, null, null);
+
+    editorManager.removeAllEditorsForPath(path);
```

Comment:
Although I do not like this option but you could if you like (not this patch) invert the direction, i.e you attach an activity listener from the editorManager and annotationManager to the LocalFilesysteModificationHandler and react on those changes.

The reason I do not like it is because you could create an anonymous producer and so not ever component that fires activities can be observerd.

Reply:
I am not convinced that such an approach would really improve the code. But I did move all the stuff dealing with the Saros state for a file (e.g. the editor pool and annotations) to separate methods. This makes it easier to uniformly adjust the logic once such an adjustment is necessary. Furthermore, it makes it easier to extract the logic if we were to decide to do so at some point in the future.